GraphConsoleAppV3/GraphRequests.cs

```diff
@@ -0,0 +1,1551 @@
+            try
+            {
+                devices = await client.Devices.ExecuteAsync();
+            }
+            catch (Exception e)
+            {
+                Program.WriteError("/nError getting devices {0}", Program.ExtractErrorMessage(e));
+            }
+
+            if (devices != null)
+            {
+                do
+                {
+                    List<IDevice> devicesList = devices.CurrentPage.ToList();
+                    foreach (IDevice device in devicesList)
+                    {
+                        if (device.ObjectId != null)
+                        {
+                            Console.WriteLine("Device ID: {0}, Type: {1}", device.DeviceId, device.DeviceOSType);
+                            IPagedCollection<IDirectoryObject> registeredOwners = device.RegisteredOwners;
+                            if (registeredOwners != null)
+                            {
+                                do
+                                {
+                                    List<IDirectoryObject> registeredOwnersList = registeredOwners.CurrentPage.ToList();
+                                    foreach (IDirectoryObject owner in registeredOwnersList)
+                                    {
+                                        Console.WriteLine("Device Owner ID: " + owner.ObjectId);
+                                    }
+                                    registeredOwners = await registeredOwners.GetNextPageAsync();
+                                } while (registeredOwners != null);
+                            }
+                        }
+                    }
+                    devices = await devices.GetNextPageAsync();
+                } while (devices != null);
+            }
+
+            #endregion
+        }
+
+        private static async Task<OAuth2PermissionGrant> CreateOAuth2Permission(
+            IActiveDirectoryClient client,
+            ServicePrincipal newServicePrincipal,
+            string graphAppObjectId)
+        {
+            #region Create a new consentable OAuth2 permission
+
+            //*********************************************************************************************
+            // Create new  oauth2 permission object
+            //*********************************************************************************************
+            if (newServicePrincipal.ObjectId != null)
+            {
+                OAuth2PermissionGrant permissionObject = new OAuth2PermissionGrant
+                {
+                    ConsentType = "AllPrincipals",
+                    Scope = "user_impersonation",
+                    StartTime = DateTime.Now,
+                    ExpiryTime = (DateTime.Now).AddMonths(12),
+                    ResourceId = graphAppObjectId,
+                    ClientId = newServicePrincipal.ObjectId
+                };
+
+                // AppId of AzureAd (Graph API)
+
+                //ClientId = objectId of servicePrincipal
+

```

Comment:
removed

---
In reply to: [81235712](https://github.com/Azure-Samples/active-directory-dotnet-graphapi-console/pull/37#discussion_r81235712) [](ancestors = 81235712)